internal/graph/lookupsubjects_test.go

```diff
@@ -0,0 +1,214 @@
+package graph
+
+import (
+	"sort"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"golang.org/x/exp/slices"
+
+	v1 "github.com/authzed/spicedb/pkg/proto/dispatch/v1"
+)
+
+func fsubs(subjectIDs ...string) *v1.FoundSubjects {
+	subs := make([]*v1.FoundSubject, 0, len(subjectIDs))
+	for _, subjectID := range subjectIDs {
+		subs = append(subs, fs(subjectID))
+	}
+	return &v1.FoundSubjects{
+		FoundSubjects: subs,
+	}
+}
+
+func fs(subjectID string) *v1.FoundSubject {
+	return &v1.FoundSubject{
+		SubjectId: subjectID,
+	}
+}
+
+func TestCreateFilteredAndLimitedResponse(t *testing.T) {
+	tcs := []struct {
+		name            string
+		subjectIDCursor string
+		input           map[string]*v1.FoundSubjects
+		limit           uint32
+		expected        map[string]*v1.FoundSubjects
+	}{
+		{
+			"basic limit, no filtering",
+			"",
+			map[string]*v1.FoundSubjects{
+				"foo": fsubs("a", "b", "c"),
+				"bar": fsubs("a", "b", "d"),
+			},
+			3,
+			map[string]*v1.FoundSubjects{
+				"foo": fsubs("a", "b", "c"),
+				"bar": fsubs("a", "b"),
+			},
+		},
+		{
+			"basic limit removes key",
+			"",
+			map[string]*v1.FoundSubjects{
+				"foo": fsubs("a", "b", "c"),
+				"bar": fsubs("b", "d"),
+			},
+			1,
+			map[string]*v1.FoundSubjects{
+				"foo": fsubs("a"),
+			},
+		},
+		{
+			"limit maintains wildcard",
+			"",
+			map[string]*v1.FoundSubjects{
+				"foo": fsubs("a", "b", "c"),
+				"bar": fsubs("b", "d", "*"),
+			},
+			1,
+			map[string]*v1.FoundSubjects{
+				"foo": fsubs("a"),
+				"bar": fsubs("*"),
+			},
+		},
+		{
+			"basic limit, with filtering",
+			"a",
+			map[string]*v1.FoundSubjects{
+				"foo": fsubs("a", "b", "c"),
+				"bar": fsubs("a", "b", "d"),
+			},
+			2,
+			map[string]*v1.FoundSubjects{
+				"foo": fsubs("b", "c"),
+				"bar": fsubs("b"),
+			},
+		},
+		{
+			"basic limit, with filtering includes both",
+			"a",
+			map[string]*v1.FoundSubjects{
+				"foo": fsubs("a", "b", "c"),
+				"bar": fsubs("a", "b", "d"),
+			},
+			3,
+			map[string]*v1.FoundSubjects{
+				"foo": fsubs("b", "c"),
+				"bar": fsubs("b", "d"),
+			},
+		},
+		{
+			"filtered limit maintains wildcard",
+			"z",
+			map[string]*v1.FoundSubjects{
+				"foo": fsubs("a", "b", "*", "c"),
+				"bar": fsubs("b", "d", "*"),
+			},
+			10,
+			map[string]*v1.FoundSubjects{
+				"foo": fsubs("*"),
+				"bar": fsubs("*"),
+			},
+		},
+	}
+
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			limits := newLimitTracker(tc.limit)
+
+			var cursor *v1.Cursor
+			if tc.subjectIDCursor != "" {
+				cursor = &v1.Cursor{
+					DispatchVersion: lsDispatchVersion,
```

Comment:
also out of scope, but for discussion: how can we automate detecting breaking cursor changes?

Reply:
We'd have to define versions via a registration system and have it automatically use the correct deserialization, returning partial information where applicable